crypto/evp/evp_enc.c

```diff
@@ -336,7 +302,25 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
             return 0;
         }
     }
+
+    /* Past this point inl has to be in bytes. */
     i = ctx->buf_len;
+    /*
+     * Aliasing in and out in this way is counter intuitive and only supported
+     * for legacy applications.
+     * In fact the check below excludes just one special case of partially
+     * overlapping buffers, while the docs say we return an error if in and out
+     * are partially overlapping.
+     * It is quite possible that the check below will be hardened in the future
+     * to return an error whenever there is any overlapping between in and out
+     * buffers.
+     */
+    if (!is_pointer_offset_by_bias(in, out, i)
+            && is_any_overlapping(out, inl + (bl - 1), in, inl)) {
```

Comment:
As mentioned earlier, I'm not really happy about is_pointer_offset_by_bias (it's unnecessary hard to audit), but in combination with second check it becomes extra tricky. I mean first it's 'in, out' and then 'out, in', br-r-r-r... As for whether or not it should be `inl + bl - 1` or something else. I'd say that *formally* speaking "something else"... You've mentioned floor_to_blocksize(inl + i) earlier... [Just in case, note there is block_mask that can be used with "floor".]